### PR TITLE
Replace bitset with std math/big

### PR DESCRIPTION
**Pull request by zhsj**

### CODE REVIEW

**Comment by rhatdan** [comment]
@kolyshkin @thaJeztah PTAL

File: go-selinux/selinux_linux.go [hunk 1]
```diff
@@ -646,7 +643,10 @@ func calculateGlbLub(sourceRange, targetRange string) (string, error) {
 		return "", ErrIncomparable
 	}
 
-	outrange := &mlsRange{low: &level{}, high: &level{}}
+	outrange := &mlsRange{
+		low:  &level{cats: new(big.Int)},
+		high: &level{cats: new(big.Int)},
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply go-selinux/selinux_linux.go:648]: Ah, those are pointers... Anyway, it's better initialize when needed (see below).

File: go-selinux/selinux_linux.go [hunk 2]
```diff
@@ -591,13 +588,13 @@ func (l1 *level) equal(l2 *level) bool {
...
 	}
-	return l1.cats.Equal(l2.cats)
+	return l1.cats.Cmp(l2.cats) == 0
 }
 
 // String returns an mlsRange as a string.
 func (m mlsRange) String() string {
 	low := "s" + strconv.Itoa(int(m.low.sens))
-	if m.low.cats != nil && m.low.cats.Count() > 0 {
+	if m.low.cats != nil && m.low.cats.BitLen() > 0 {
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply go-selinux/selinux_linux.go:597]: OTOH you don't really need pop count here! You need to know if the value is non-zero, and `BitLen()` actually answers that!

File: go-selinux/selinux_linux.go [hunk 3]
```diff
@@ -548,39 +548,36 @@ func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
...
+			str += ",c" + strconv.FormatInt(int64(i), 10)
+		} else if length > 1 {
+			str += ".c" + strconv.FormatInt(int64(i), 10)
 		}
-		len = 0
-		i = next
+		i++
+		length = 0
 	}
-
```
Comment (kolyshkin) [inline go-selinux/selinux_linux.go:583]: please keep this empty line

**Review by kolyshkin — CHANGES_REQUESTED** [review]
Nice work, LGTM except a nit about initialization

File: go-selinux/selinux_linux.go (hunk 3 above)
Comment (zhsj) [reply go-selinux/selinux_linux.go:583]: fixed

File: go-selinux/selinux_linux.go [hunk 4]
```diff
@@ -656,10 +656,10 @@ func calculateGlbLub(sourceRange, targetRange string) (string, error) {
 
 	/* find the intersecting categories */
 	if s.low.cats != nil && t.low.cats != nil {
-		outrange.low.cats = s.low.cats.Intersection(t.low.cats)
+		outrange.low.cats.And(s.low.cats, t.low.cats)
```
[the remark this answers is not in the document]
Comment (zhsj) [reply go-selinux/selinux_linux.go]: done

[the remark this answers is not in the document]
Comment (zhsj) [reply go-selinux/selinux_linux.go]: ok, this leaves cats nil, and needs to check if it's nil in other functions. What's your preference?

[the remark this answers is not in the document]
Comment (zhsj) [reply go-selinux/selinux_linux.go]: never mind, it's already check elsewhere, just need one more check.

**Comment by thaJeztah** [comment]
looks like this was updated; @kolyshkin ptal if your comments were addressed 🤗

File: go-selinux/selinux_linux.go [hunk 5]
```diff
@@ -548,37 +548,35 @@ func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
...
+		if c.Bit(i) == 0 {
+			i++
+			continue
+		}
+		if length == 0 {
 			if str != "" {
 				str += ","
 			}
-			str += "c" + strconv.Itoa(int(i))
+			str += "c" + strconv.FormatInt(int64(i), 10)
```
Comment (kolyshkin) [inline go-selinux/selinux_linux.go:565]: From the docs:

> Itoa is equivalent to FormatInt(int64(i), 10).

so why change this?

File: go-selinux/selinux_linux.go [hunk 6]
```diff
@@ -548,37 +548,35 @@ func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
...
+			i++
+			length++
+			continue
 		}
-		if !e {
-			break
+		if length == 1 {
+			str += ",c" + strconv.FormatInt(int64(i), 10)
+		} else if length > 1 {
+			str += ".c" + strconv.FormatInt(int64(i), 10)
```
Comment (kolyshkin) [inline go-selinux/selinux_linux.go:576]: Same here, `strconv.Itoa(i)` seems easier (feel free to ignore though).

File: go-selinux/selinux_linux.go (hunk 5 above)
Comment (zhsj) [reply go-selinux/selinux_linux.go:565]: no particular reason,  probably i deleted the old lines when rewrote and used my own tastes.

grep the strconv usage in package, seems iota and itoa are more used. let's keep the style.

**Review by kolyshkin — DISMISSED** [review]
LGTM

@gitstashpop PTAL

**Comment by kolyshkin** [comment]
LGTM, but would like @gitstashpop to take a look before merging this.

**Review by kolyshkin — APPROVED** [review]
LGTM, thanks!

File: go-selinux/selinux_linux.go [hunk 7]
```diff
@@ -548,37 +548,35 @@ func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
...
+func bitsetToStr(c *big.Int) string {
 	var str string
-	i, e := c.NextSet(0)
-	len := 0
-	for e {
-		if len == 0 {
+
+	i := int(c.TrailingZeroBits())
+	length := 0
+	for i < c.BitLen() {
```
Comment (gitstashpop) [inline go-selinux/selinux_linux.go:556]: nit: this could be a `for init; condition; increment` loop

**Review by gitstashpop — APPROVED** [review]
changes lgtm, just a nit about for loop style

File: go-selinux/selinux_linux.go (hunk 7 above)
Comment (zhsj) [reply go-selinux/selinux_linux.go]: done

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@kolyshkin @gitstashpop ptal (change was pushed and made your last review "outdated")

**Review by gitstashpop — APPROVED** [review]
lgtm

**Comment by rhatdan** [comment]
Thanks @zhsj

**Pull request merged**